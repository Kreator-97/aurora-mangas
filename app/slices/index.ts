export { default as uiSlice } from './uiSlice'
export { default as authSlice } from './authSlice'
export { default as shoppingCartSlice } from './shoppingCartSlice'
