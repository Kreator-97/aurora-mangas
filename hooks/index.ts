export * from './useForm'
export * from './useCounter'
