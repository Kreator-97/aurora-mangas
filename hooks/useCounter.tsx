import { useState } from 'react'

interface Props {
  initial?: number;
  minValue?: number;
}

export const useCounter = ({initial = 0, minValue }:Props) => {
  const [ counter, setCounter] = useState(initial)

  const increment = (factor:number) => {
    setCounter(counter + factor)
  }

  const decrement = (factor:number) => {
    if ( minValue !== undefined ) {
      const max = Math.max(minValue, counter-factor)
      setCounter(max)
      return
    }
    setCounter(counter - factor)
  }

  const reset = () => setCounter(initial)

  return { increment, decrement, reset, counter }
}
