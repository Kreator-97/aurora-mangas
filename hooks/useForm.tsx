import { ChangeEvent, useState } from 'react'

export function useForm<T> (initialValue: T) {
  const [ formState, setFormState ] = useState<T>(initialValue)
  
  const onInputChange = ({target}:ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    const {name, value} = target
    setFormState({
      ...formState,
      [name]: value,
    })
  }

  const onResetForm = () => {
    setFormState(initialValue)
  }

  return {
    ...formState,
    formState,
    onInputChange,
    onResetForm
  }
}
