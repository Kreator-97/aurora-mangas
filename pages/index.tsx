import type { GetServerSideProps, NextPage } from 'next'

import { CardGrid, CardManga, Pagination } from '../components'
import { dbSeries } from '../database'
import { Serie } from '../interfaces'
import { AppLayout } from '../layouts'

interface Props {
  series: Serie[]
}

const Home: NextPage<Props> = ({series}) => {

  return (
    <AppLayout title='Aurora Mangas | Página de inicio'>
      <h1 className='title'>
        Series nuevas
      </h1>
      <CardGrid series={ series } />
      <h1 className='title'>
        Últimos lanzamientos 
      </h1>

      <CardGrid gridCols='grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4' >
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
        <CardManga />
      </CardGrid>
      <Pagination totalPages={20}/>
    </AppLayout>
  )
}

export default Home

export const getServerSideProps: GetServerSideProps = async (ctx) => {
  const series = await dbSeries.getSeries()

  return {
    props: {
      series,
    }
  }
}
