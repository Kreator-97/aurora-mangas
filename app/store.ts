import { configureStore } from '@reduxjs/toolkit'
import { uiSlice, authSlice, shoppingCartSlice } from './slices'

export const store = configureStore({
  reducer: {
    ui: uiSlice,
    auth: authSlice,
    cart: shoppingCartSlice
  },
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
