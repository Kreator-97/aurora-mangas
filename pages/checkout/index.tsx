import { GetServerSideProps, NextPage } from 'next'
import { Session } from 'next-auth'
import { getSession } from 'next-auth/react'
import Image from 'next/image'

import { AppLayout } from '../../layouts'
import { dbLocal, formatPrice } from '../../util'
import { removeItem, setAmount } from '../../app/slices/shoppingCartSlice'
import { SelectAmount } from '../../components/selectAmount/SelectAmount'
import { useAppDispatch, useAppSelector } from '../../app/hooks'

interface Props {
  user: Session | null;
}

const CheckoutPage:NextPage<Props> = ({user}) => {
  const dispatch = useAppDispatch()
  const { items, total } = useAppSelector(state => state.cart)

  const onIncrement = (amount:number, id:string ) => {
    const newAmount = amount < 1 ? 1 : amount
    dispatch(setAmount({amount: newAmount, id}))
    dbLocal.setProductAmountInLocal({items, total}, newAmount, id)
  }
  
  const onDecrement = (amount:number, id:string ) => {
    const newAmount = amount < 1 ? 1 : amount
    dispatch(setAmount({amount: amount < 1 ? 1 : amount, id}))
    dbLocal.setProductAmountInLocal({items, total}, newAmount, id)
  }

  const onRemoveItem = (id:string) => {
    dispatch(removeItem({id}))
    dbLocal.removeItemInLocal({items, total}, id)
  }

  if ( items.length === 0) {
    return (
      <AppLayout title="Resumen de compra">
        <div className='mb-4'>
          <h1 className='title'>Resumen de compra</h1>
          <h2 className='text-center'>Este es el resumen de compra, si está satisfecho confirme su pedido</h2>
        </div>
        <h2 className='text-center text-lg'>No tienes elementos en el carrito de compras</h2>
      </AppLayout>
    )
  }

  return (
    <AppLayout title="Resumen de compra">
      <div className='mb-4'>
        <h1 className='title'>Resumen de compra</h1>
        <h2 className='text-center'>Este es el resumen de compra, si está satisfecho confirme su pedido</h2>
      </div>

      <div className='max-w-sm grid grid-cols-1 gap-4 mx-auto mb-8'>
        {
          items.map( item => {
            const title = `${item.product.serie.name} #${item.product.number}`
            return (
              <div key={item.product.id} className="flex gap-4 justify-center">
                <Image
                  src={item.product.imgURL}
                  layout="fixed"
                  width={100}
                  height={150}
                />
                <div>
                  <h2 className='text-xl '>{title}</h2>
                  <h2 className='text-xl '>Subtotal: { formatPrice( item.amount * item.product.price )}</h2>
                  <button onClick={ () => onRemoveItem(item.product.id) } className='text-error text-lg'>Eliminar</button>
                  <SelectAmount
                    initial={ item.amount }
                    onIncrement={ onIncrement }
                    onDecrement={ onDecrement }
                    id={item.product.id}
                  />
                </div>
              </div>
            )
          })
        }
      </div>
      <div className='max-w-sm grid grid-cols-1 gap-4 mx-auto mb-4'>
        { 
          !user && (
            <div>
              <p className='text-center text-alert'>Necesitas iniciar sesión para realizar tu compra</p>
              <p className='text-center text-alert'>Inicia sesión o crea una cuenta para continuar</p>
            </div>
          )
        }
        <h2 className='title'>Total: {formatPrice(total)}</h2>
        <button
          className='btn bg-accent w-full'
          disabled={ !user }
        >
          Confirmar pedido
        </button>
        <p className='text-success text-center' >El pago es procesado a través de Paypal</p>
      </div>
    </AppLayout>
  )
}

export default CheckoutPage

export const getServerSideProps:GetServerSideProps = async (ctx) => {
  const session = await getSession(ctx)

  return {
    props: { user: session }
  }
}
