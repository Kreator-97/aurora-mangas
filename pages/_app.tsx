import type { AppProps } from 'next/app'
import { Provider } from 'react-redux'
import { SessionProvider } from 'next-auth/react'
import { ApolloProvider } from '@apollo/client'

import { client } from '../graphql/client'
import { store } from '../app/store'
import '../styles/globals.css'
import 'animate.css'

function MyApp({ Component, pageProps }: AppProps) {
  return (
    <SessionProvider>
      <ApolloProvider client={ client }>
        <Provider store={ store }>
          <Component {...pageProps } />
        </Provider>
      </ApolloProvider>
    </SessionProvider>
  ) 
}

export default MyApp
