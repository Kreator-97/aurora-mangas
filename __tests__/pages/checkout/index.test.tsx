import { render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import { store } from '../../../app/store'
import { default as CheckoutPage } from '../../../pages/checkout'
import { mangas, users } from '../../fixtures/db'

// mocks
import { useSession } from 'next-auth/react'
import { useAppSelector } from '../../../app/hooks'

jest.mock('next-auth/react')
jest.mock('next/router', () => {
  return {
    useRouter: () => ({ query: { q: '' } })
  }
})
jest.mock('../../../components/sidebar/Sidebar')
jest.mock('../../../app/hooks')

describe('tests on Checkout Page', () => {
  (useSession as jest.Mock).mockReturnValue(() => {
    return {
      status: 'authenticated',
      data: users[0],
    }
  })

  test('should to match with snapshot', () => {
    (useAppSelector as jest.Mock).mockReturnValue(({
      cart: { items: [{ product: mangas[0], amount: 1 }], total: mangas[0].price }
    }).cart)

    const session = {
      status : 'authenticated',
      user : users[0],
      expires: '0'
    }

    const { container } = render(
      <Provider store={store} >
        <CheckoutPage user={ session }/>
      </Provider>
    )
    expect(container).toMatchSnapshot()
  })

  test('should to show message when there are not items on cart', () => {
    (useAppSelector as jest.Mock).mockReturnValue(({
      cart: {
        items: [],
        total: 0
      },
    }).cart)

    const { container } = render(
      <Provider store={store}>
        <CheckoutPage user={ null }/>
      </Provider>
    )
    // expect(container).toMatchSnapshot()
    const message = screen.getByText('No tienes elementos en el carrito de compras')
    expect(message).toBeInTheDocument
  })

  test('should to show login button when user is not authenticated', () => {
    (useAppSelector as jest.Mock).mockReturnValue(({
      cart: { items: [{ product: mangas[0], amount: 1 }], total: mangas[0].price }
    }).cart)

    render(
      <Provider store={store}>
        <CheckoutPage user={ null }/>
      </Provider>
    )
    
    expect( screen.getByText('Iniciar sesiÃ³n') ).toBeInTheDocument
  })

  test('"Confirmar pedido" button should to be active when user is logged and exists items on cart', () => {

    (useAppSelector as jest.Mock).mockReturnValue(({
      cart: { items: [{ product: mangas[0], amount: 1 }], total: mangas[0].price }
    }).cart)

    const session = {
      status : 'authenticated',
      user : users[0],
      expires: '0'
    }

    const { container } = render(
      <Provider store={store} >
        <CheckoutPage user={ session }/>
      </Provider>
    )

    const confirm = screen.getByText('Confirmar pedido')
    expect(confirm.getAttributeNames()).not.toContain('disabled')
  })
})
