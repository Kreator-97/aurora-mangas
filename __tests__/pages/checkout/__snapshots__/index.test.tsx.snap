// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on Checkout Page should to match with snapshot 1`] = `
<div>
  <nav
    class="container"
  >
    <div
      class="content"
    >
      <svg
        color="white"
        cursor="pointer"
        data-testid="menu"
        fill="none"
        height="32"
        stroke="currentColor"
        stroke-width="0.5"
        style="color: white;"
        viewBox="0 0 24 24"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 6h16M4 12h16M4 18h16"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 150px; height: 30px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <img
          alt="Aurora Mangas Logo"
          class="cursor-pointer"
          data-nimg="fixed"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
      <div
        class="relative cursor-pointer"
        data-testid="cart-menu"
      >
        <svg
          color="white"
          cursor="pointer"
          fill="none"
          height="32"
          stroke="currentColor"
          stroke-width="0"
          style="color: white;"
          viewBox="0 0 24 24"
          width="32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 3h2l.4 2M7 13h10l4-8H5.4M7 13L5.4 5M7 13l-2.293 2.293c-.63.63-.184 1.707.707 1.707H17m0 0a2 2 0 100 4 2 2 0 000-4zm-8 2a2 2 0 11-4 0 2 2 0 014 0z"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="w-6 h-6 text-center font-semibold bg-accentDark bg-opacity-95 rounded-full absolute -top-2 -right-2"
        >
          1
        </span>
      </div>
      <form
        class="flex col-span-2"
        data-testid="search-form"
      >
        <input
          class="w-full rounded border-cyan-400 border p-1 text-sm text-dark"
          placeholder="Encuentra tu manga favorito. Ejemplo: one punch man #1"
          type="search"
          value=""
        />
      </form>
      <svg
        color="white"
        cursor="pointer"
        fill="none"
        height="32"
        stroke="currentColor"
        stroke-width="0"
        style="color: white;"
        viewBox="0 0 24 24"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 16l2.879-2.879m0 0a3 3 0 104.243-4.242 3 3 0 00-4.243 4.242zM21 12a9 9 0 11-18 0 9 9 0 0118 0z"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </div>
  </nav>
  <div
    class="fixed top-0 right-0 min-h-screen h-screen min-w-[360px] bg-dark z-50 transition-transform translate-x-full"
  >
    <div
      class="flex justify-between items-center border-b border-strokeLight p-4"
      style="height: 65px;"
    >
      <svg
        cursor="pointer"
        data-testid="close-shopping-cart-icon"
        fill="currentColor"
        height="28"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="28"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
      <h2
        class="text-center text-2xl"
      >
        Carrito de compra
      </h2>
      <div
        class="w-7"
      />
    </div>
    <div
      class="flex flex-col gap-4 items-center overflow-y-scroll cartlist-items"
      style="padding-top: 1rem; padding-bottom: 1rem;"
    >
      <div
        class="flex"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 100px; height: 100px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
        >
          <img
            alt="One punch man no.12"
            class="object-contain"
            data-nimg="fixed"
            data-testid="cart-item"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
        <div
          class="px-2 w-52"
        >
          <h2>
            One Punch Man #12
          </h2>
          <h2>
            Cantidad: 
            1
          </h2>
          <h2>
            Subtotal: 
            $99.00
          </h2>
          <button
            class="text-error text-sm"
          >
            Eliminar del carrito
          </button>
        </div>
      </div>
    </div>
    <div
      class="bg-dark absolute left-0 w-full px-4 py-4 border-t border-strokeLight"
      style="height: 108px;"
    >
      <h2
        class="text-center text-2xl mb-2"
      >
        Total global: 
        $99.00
      </h2>
      <button
        class="btn bg-accent w-full"
      >
        Ir al resumen
      </button>
    </div>
  </div>
  <main
    class="main"
    style="max-width: var(--xl);"
  >
    <div
      class="mb-4"
    >
      <h1
        class="title"
      >
        Resumen de compra
      </h1>
      <h2
        class="text-center"
      >
        Este es el resumen de compra, si está satisfecho confirme su pedido
      </h2>
    </div>
    <div
      class="max-w-sm grid grid-cols-1 gap-4 mx-auto mb-8"
    >
      <div
        class="flex gap-4 justify-center"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 100px; height: 150px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
        >
          <img
            alt="One punch man no.12"
            data-nimg="fixed"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
        <div>
          <h2
            class="text-xl "
          >
            One Punch Man #12
          </h2>
          <h2
            class="text-xl "
          >
            Subtotal: 
            $99.00
          </h2>
          <button
            class="text-error text-lg"
          >
            Eliminar
          </button>
          <div>
            <h2
              class="text-lg mb-2 text-center"
            >
              Seleccione la cantidad
            </h2>
            <div
              class="flex gap-4 mb-2 justify-center"
            >
              <div>
                <svg
                  cursor="pointer"
                  data-testid="decrement-icon"
                  fill="none"
                  height="28"
                  stroke="currentColor"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="28"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15 12H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
              <div
                class="font-medium text-2xl px-2"
                data-testid="amount-box"
              >
                1
              </div>
              <div>
                <svg
                  cursor="pointer"
                  data-testid="increment-icon"
                  fill="none"
                  height="28"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 24 24"
                  width="28"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 9v3m0 0v3m0-3h3m-3 0H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="max-w-sm grid grid-cols-1 gap-4 mx-auto mb-4"
    >
      <h2
        class="title"
      >
        Total: 
        $99.00
      </h2>
      <button
        class="btn bg-accent w-full"
      >
        Confirmar pedido
      </button>
      <p
        class="text-success text-center"
      >
        El pago es procesado a través de Paypal
      </p>
    </div>
    <div
      style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
    />
  </main>
  <footer
    class="min-h-[150px] border-t border-solid border-stroke"
  >
    <h2
      class="text-xl md:text-3xl text-center p-4 mt-8"
    >
      Síguenos en nuestras redes sociales
    </h2>
    <div
      class="flex justify-center gap-8"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 32px; height: 32px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <img
          alt="facebook"
          data-nimg="fixed"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; width: 36px; height: 36px; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative;"
      >
        <img
          alt="twitter"
          data-nimg="fixed"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
  </footer>
</div>
`;
